src/linear_algebra/free_module_pid.lean

```diff
@@ -404,6 +659,129 @@ noncomputable def module.free_of_finite_type_torsion_free' [module.finite R M]
   Σ (n : ℕ), basis (fin n) R M :=
 module.free_of_finite_type_torsion_free module.finite.exists_fin.some_spec.some_spec
 
+section smith_normal
+
+/-- A Smith normal form basis for a submodule `N` of a module `M` consists of bases for `M` and `N`
+such that the inclusion map `N → M` can be written as a diagonal matrix in Smith normal form. -/
+@[nolint has_inhabited_instance]
+structure basis.smith_normal_form (N : submodule R M) (ι : Type*) (n : ℕ) :=
+(bM : basis ι R M) (bN : basis (fin n) R N) (f : fin n ↪ ι)
+(a : fin n → R) (snf : ∀ i, (bN i : M) = a i • bM (f i))
```

Comment:
You are doing something a bit more general right? Because you include the `n : nat` that allows you to do "partial(?)" SNF bases?

Unrelated: I think I would prefer having one structure field per line.

Reply:
The point of using two different indices is that submodules that are not of full rank can still have SNF bases. That is, the matrix you end up is rectangular with `A i (f i) = a i` and `A i j = 0` otherwise. Or you can view it as a full `ι × ι` matrix with `A (f i) (f i) = a i` and `A i j = 0` otherwise, including on the diagonal outside of the image of `f`. And the reason for using `fin n` rather than an abstract type `κ` is that you can ensure `a i | a (i + 1)`.

Maybe "diagonal matrix" is a poor choice of words, I have changed it into "a (rectangular) matrix with `a` along the diagonal: in Smith normal form."

I have also reformatted the fields.